shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -556,24 +556,51 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
   def computeVariances(tyDefs: List[TypeDef], ctx: Ctx): Unit = {
     var varianceUpdated: Bool = false;
 
-    def fieldVarianceHelper(fieldTy: FieldType, curVariance: VarianceInfo)(implicit tyDef: TypeDef): Unit = {
+    def fieldVarianceHelper(fieldTy: FieldType, curVariance: VarianceInfo)(implicit tyDef: TypeDef, visited: MutSet[TypeVariable]): Unit = {
         fieldTy.lb.foreach(lb => updateVariance(lb, curVariance.flip))
         updateVariance(fieldTy.ub, curVariance)
     }
 
-    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit tyDef: TypeDef): Unit =
+    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit tyDef: TypeDef, visited: MutSet[TypeVariable]): Unit =
       trace(s"updateVariance(ty: $ty of ${ty.getClass()}, curVariance: $curVariance)(${tyDef.nme.name}.${tyDef.tvarVariances})") {
         ty match {
           case ProxyType(underlying) => updateVariance(underlying, curVariance)
           case TraitTag(_) | ClassTag(_, _) => ()
           case ExtrType(pol) => ()
           case t: TypeVariable =>
-            // a type variable is bivariant when we know nothing about it
-            val oldVariance = tyDef.tvarVariances.getOrElseUpdate(t, VarianceInfo.bi)
-            val newVariance = oldVariance && curVariance
-            if (newVariance =/= oldVariance) {
-              tyDef.tvarVariances(t) = newVariance
-              varianceUpdated = true
+            // if a type variable has not been visited before
+            // try visiting it's upper or lower bounds to
+            // handle recursion cases as well
+            if (!visited.contains(t)) {
+              visited.add(t)
+              // handle cases where there is no recursive
+              // definition by recomputing udpateVariance for
+              // the type variable but this time it's been added
+              // to the visited set
+              curVariance match {
+                case VarianceInfo(true, true) | VarianceInfo(false, true) =>
+                  if (t.upperBounds.isEmpty)
+                    updateVariance(t, curVariance)
+                  else
+                    t.upperBounds.foreach(ub => updateVariance(ub, curVariance))
+                case VarianceInfo(true, false) =>
+                  if (t.lowerBounds.isEmpty)
+                    updateVariance(t, curVariance)
+                  else
+                    t.lowerBounds.foreach(lb => updateVariance(lb, curVariance))
+                case VarianceInfo(false, false) => updateVariance(t, curVariance)
+              }
+            }
```

Comment:
This logic is very wrong.

There should be no special-casing when the bounds are empty.

You should test for `visited.contains(pol -> t)` where `pol` depends on `curVariance`, as explained before.